### PR TITLE
ARCore SDK Preview 2 / 2017.3

### PR DESCRIPTION
**Pull request by stella3d**
Saw #17, also for updating ARCore, as i came to open this PR!  

Was happy to see it, but i would like to instead propose a simpler, more focused update for now that should allow applications built using the first version of ARInterface to run on ARCore Preview 2 as well, with no API changes.  

I see the other PR deals with resume functionality, too.  I think we should integrate that separately from the main dependency upgrade.

#### some notes:
1) in `StartService`, i'm not waiting for the service to start (because that's now async), but this hasn't been a problem in my testing.  Changing that will require changing the API.
2) instead of managing the camera ourselves, we pass that off to `ARCoreBackgroundRenderer`, which seems to work great.
3) I added a copy of the `ARBackground` material to `Resources/Materials`, which is used by `ARCoreInterface`
4) ARCore preview 2 requires 2017.3.0f2, so i updated the ProjectVersion to that (which the readme already recommended)
5) the ARKit `InstancedSurfaceShader.shader` & `PostProcessing/Editor/Utils/CurveEditor.cs` both trigger the script upgrader in 2017.3, so i committed the upgraded versions.

I've built a number of the sample apps onto my Pixel and they've been working as expected, but i haven't tried all of them yet.  Please feel free to come by my desk with any questions , or if you want to see it working on-device.  

Implementing this made think we should try to find a better way of visualizing planes for ARCore - this scaled prefab looks off from a distance, even before this update.  only up close or well aligned does the prefab look right.  

In addition, we should probably start creating tagged releases - one for the old version, one for the new, if this is merged.

### CODE REVIEW

**Comment by stella3d** [comment]
this should also fix issue #7

**Comment by iBicha** [comment]
Haha this is awesome. 
Just to clarify, #17 only made changes that would allow the interface usage and behaviour to be the same between the old and new SDK, meaning the demos will all play the same with zero changes, with one exception, that `TryGetCameraImage` would fail (I'll get to that in a bit). So having #17 or #18 would probably have existing projects running without any change.

For the `TryGetCameraImage` I'll just leave this here, for anyone wondering why it's not working for either PRs:

In SDK preview 1, the API returned buffers directly from the camera, that you can manipulate directly, which was fine.
With SDK preview 2, the API returned a texture. It sounds better at first, but here are a few gotchas:

1. The texture can be null at some situations
2. The texture will have width and height equal to zero, since this information is not available. (See `FrameManager.cs#141`)
3. This texture is externally created ([OES_EGL_image_external](https://www.khronos.org/registry/OpenGL/extensions/OES/OES_EGL_image_external.txt)) which means it cannot be read through `GetPixels` or `GetRawTextureData` making it useless if you need to use it for CV.
4. The texture will not render using any material! it has to be rendered using a shader with a `samplerExternalOES` sampler (which is what the ARBackground material is using)

I wanted to implement something that kept the existing API to get the camera image (yuv data) so I followed this approach:

1. Created a `RenderTexture`, and rendered the texture from the frame to it using the `ARBackground` shader (`Graphics.Blit`)
2. Set the render texture to active, and used `ReadPixels` to copy pixels from the `RenderTexture`, to another `Texture2D` that I created.
3. Get the raw data using `GetRawTextureData` (since the texture I created was readable)
4. Convert the resulting buffer from `ARGB32` to `YUV2` color space.

The approach was successful, but because there is a lot of copying around of data, also because the CPU is basically waiting for the GPU to finish copying in `ReadPixels`, and finally because of the color conversion, the result was really slow for me, so I dropped the idea of keeping the same Interface to get the image data. `TryGetCameraImage` will have to change. ¯\_(ツ)_/¯

**tl;dr**
**If you really need to use the image pixels**, your two options are:
1-Stick to the first SDK preview
2-Follow the approach in the `ComputerVision` example, where reading pixels is done in a native c plugin, and the images are 1 frame old, which is to give the GPU time to copy it to a buffer before the CPU access it (So, you will notice some lag, even in the `ComputerVision` example, it is double buffered and will always get the the data a frame later after your request)

**Comment by Crashthatch** [comment]
I have tested this PR on Android and it mostly works as expected for me.

Your point 1, I'm not sure I agree with the change to make [ArCoreInterface.StartService()](https://github.com/Unity-Technologies/experimental-ARInterface/pull/18/files#diff-551c80bf137dd5ac20e1f204643f8c3fL51) always return true. 

I assume the benefit of doing it this way is that it fixes #11, and the video appears on first grant (rather than requiring a second call to re-enable the ARController once permission is granted, as in preview-1)?

However, in fixing this, it makes it impossible to wait until camera permission has been granted to take another action (either by callback or by polling). My app needs to wait for permission to be granted & startService is complete so that I can then prompt for another (location) permission. With the change this PR makes, there is no way to tell when the camera permission has been accepted and the session established (and also no way to register for a callback).

I see there's a new ConnectServiceSync(), which might do what I want, but that's not in the ARInterface, so can't be used externally.

I think the interface should ultimately be changed (as you and #17 both suggest), but is this PR the right place to do that? In my fork of @stella3d's branch  I've reverted the "always return true" functionality and it works as before (I also use the fix from PR #12 as a workaround for #11).

-----

I also found TryGetCameraImage is not implemented and does not return an image.

@iBicha thanks for your detailed description of the camera issues and problems- I was about to try your approach and you've convinced me that would be a waste of time. I'm going to disable the CV portion of my app for now (a shame, but for me it's not worth the effort) and maybe try to reimplement once ARCore is out of developer preview (or at least someone else has solved the problem of getting a camera image without too much cpu overhead - 1 frame delay is not an issue for me) & hopefully by then there'll be a consistent interface to getCameraImage on both Android & iOS too. 

Looks like the same issue is being discussed here? https://github.com/google-ar/arcore-android-sdk/issues/92

**Comment by iBicha** [comment]
@Crashthatch I ended up implementing `TryGetCameraImage` , commit pushed to #17 
It works decently, 1 frame old

**Comment by SetoKaiba** [comment]
Where's the original author? It's inconvenient for developers to choose between forks. I think it's better merged the changes.
As I read, iBicha's implementation is with the support of IsSupported property and TryGetCameraImage while the stella3d's is with the fix for plane update and use ARCoreBackgroundRenderer instead.
They're both with feature that the other aren't. 
I would like to see a unified merge for it.

**Comment by Crashthatch** [comment]
@jimmy-jam Can you merge one or both into the official master?
